go/marshal/decode_test.go

```diff
@@ -500,8 +507,51 @@ func TestDecodeRecursive(t *testing.T) {
 
 	assert.Equal(Node{
 		1, []Node{
-			Node{2, []Node(nil)},
-			Node{3, []Node(nil)},
+			Node{2, []Node{}},
+			Node{3, []Node{}},
 		},
 	}, n)
 }
+
+func TestDecodeMap(t *testing.T) {
+	assert := assert.New(t)
+	var m map[string]int
+
+	err := Unmarshal(types.NewMap(
+		types.String("a"), types.Number(1),
+		types.String("b"), types.Number(2),
+		types.String("c"), types.Number(3)), &m)
+	assert.NoError(err)
+	assert.Equal(map[string]int{"a": 1, "b": 2, "c": 3}, m)
+
+	m = nil
+	err = Unmarshal(types.NewMap(
+		types.String("a"), types.Number(1),
```

Comment:
Is this repeated test for the caching logic or something? Seems like you should be resuing a types.Map and map[string]int instead of repeating.